sklearn/decomposition/dict_learning.py

```diff
@@ -343,35 +352,32 @@ def _update_dict(dictionary, Y, code, verbose=False, return_r2=False,
         Updated dictionary.
 
     """
-    n_components = len(code)
-    n_samples = Y.shape[0]
+    n_features, n_components = B.shape
     random_state = check_random_state(random_state)
     # Residuals, computed 'in-place' for efficiency
-    R = -np.dot(dictionary, code)
-    R += Y
+    R = -np.dot(dictionary, A)
+    R += B
     R = np.asfortranarray(R)
-    ger, = linalg.get_blas_funcs(('ger',), (dictionary, code))
+    ger, = linalg.get_blas_funcs(('ger',), (dictionary, A))
     for k in range(n_components):
         # R <- 1.0 * U_k * V_k^T + R
-        R = ger(1.0, dictionary[:, k], code[k, :], a=R, overwrite_a=True)
-        dictionary[:, k] = np.dot(R, code[k, :].T)
+        R = ger(1.0, dictionary[:, k], A[k, :], a=R, overwrite_a=True)
+        dictionary[:, k] = np.dot(R, A[k, :])
```

Comment:
Ah, looks like i actually introduced a bug: the line `dictionary[:, k] = np.dot(R, A[k, :])` shouldn't be there anymore. This additional buggy computation should account for some of the speed loss you're experiencing. Fixed.

BTW, this is what I'm doing:
```
    .5 * ||Xt - DC||_F^2 = .5 * tr(DtDA) - 2 * tr(DtB) + tr(XXt),

    as a function of the dictionary (V), where

    B = XtCt, A = CCt, R = B - DA.

    Note that the update of the jth atom is given by (see eqn 10 of
    ref paper below):

    R = R + D[:, k]A[k, :]  # rank-1 update
    D[:, k] = R[:, k] / A[k, k]  # = D[:, k] + R[k] / A[k, k]
    D[:, k] = proj(D[:, k])  # eqn 10
    R = R - D[:, k]A[k, :]  # rank-1 update
```